### PR TITLE
fix: Empty column handling improvements

### LINKED ISSUE #892: Segmentation fault when updating empty dataframe
### Describe the bug

If you update an empty dataframe with an empty dataframe, ArcticDB segfaults.

### Steps/Code to Reproduce

```python
>>> from arcticdb import Arctic
>>> ac = Arctic("lmdb:///tmp/aa")
>>> import pandas as pd 
>>> ac.create_library("aaa")
Library(Arctic(config=LMDB(path=/tmp/aa)), path=aaa, storage=lmdb_storage)
>>> lib = ac["aaa"]
>>> lib
Library(Arctic(config=LMDB(path=/tmp/aa)), path=aaa, storage=lmdb_storage)
>>> lib.write("aaa", pd.DataFrame())
VersionedItem(symbol='aaa', library='aaa', data=n/a, version=0, metadata=None, host='LMDB(path=/tmp/aa)')
>>> lib.update("aaa", pd.DataFrame())
Segmentation fault (core dumped)
```

### Expected Results

Sensible validation error about the update.

### OS, Python Version and ArcticDB Version

Python: 3.8.16 (default, Mar  5 2023, 12:11:33) 
[GCC 10.2.1 20210130 (Red Hat 10.2.1-11)]
OS: Linux-5.15.0-76-generic-x86_64-with-glibc2.17
ArcticDB: 3.0.0

### Backend storage used

LMDB

### Additional Context

_No response_

### LINKED ISSUE #940: Reading an empty series leads to a crash when as_of parameter is used
### Describe the bug

When using the as_of parameter in reads, if the return data is an empty series, this leads to a crash.
The problem seems to reproduce only on 4.0.0 and only when as_of is used.
See full repro below 

### Steps/Code to Reproduce

``` python
import pandas as pd
from arcticdb import Arctic
ac = Arctic("lmdb://tst")
ac.create_library("1")
lib = ac["1"]
s = pd.Series(dtype="datetime64[ns]")
lib.write("sym", s)
lib.read("sym", as_of=0).data
```

### Expected Results

The read operation should return an empty series.
On 3.0.0, the type of the series is: **Series([], dtype: datetime64[ns])**

### OS, Python Version and ArcticDB Version

ArcticDB: 4.0.0

### Backend storage used

Any storage

### Additional Context

_No response_

### LINKED ISSUE #987: Empty column handling improvements
Current behaviour (inferred, needs to be verified by looking at the code):

- Any columns with no values are stored as empty-type
- On read, empty-type columns are given dtype float64 for Pandas 1.X clients, and object for Pandas 2.X clients
- This leads to the inability to round-trip columns that have no values, but where the user explicitly specified a non-float64 numeric dtype

Desired behaviour:
- Columns of numeric (which includes datetimes) dtype != float64 with no values should always be stored with the provided dtype (never empty-type).
- Columns with no values of dtype float64 or object in Pandas 1.X should be stored with empty-type
- Columns with no values of dtype float64 in Pandas 2.X should be stored with with dtype float64
- Columns with no values of dtype object in Pandas 2.X should be stored with empty-type
- On read, empty-type columns should be converted to float64 for Pandas 1.X clients, and object for Pandas 2.X clients
- 
This means that the C++ layer will need to be aware of the Pandas version used to write the data.

Known limitations:

- Columns with no values of dtype object written with Pandas 1.X will be returned as type float64 (i.e. will not roundtrip correctly)
- Columns with no values and an explicitly specified dtype of float64 written with Pandas 1.X will be returned with dtype object when read with clients using Pandas 2.X
- Similarly, columns with no values and an explicitly specified dtype of objectwritten with Pandas 2.X will be returned with dtype float64 when read with clients using Pandas 1.X

```
| Reader / Writer | Pandas 1 Float | Pandas 1 Object | Pandas 2 Float | Pandas 2 Object |
| +++++++++++++++ | ++++++++++++++ | +++++++++++++++ | ++++++++++++++ | +++++++++++++++ |
| Pandas 1       || Float          | Float (!)       | Float          | Float (!)       |
| Pandas 2       || Object (OK)    | Object          | Float          | Object          |
```

### PR DESCRIPTION
**Pull request by jjerphan**
#### Reference Issues/PRs

Fix #892.
Fix #940.
Fix #987.

#### What does this implement or fix?

This PR changes the handling of empty columns as described in https://github.com/man-group/ArcticDB/issues/987#issue-1956701712, i.e. schematically:

Screenshot from 2023-11-13 08-52-31

#### Any other comments?

It also allow updating symbols with empty `DataFrame` and `Series`.

#### Checklist

<details>
  <summary>
   Checklist for code changes...
  </summary>
 
 - [x] Have you updated the relevant docstrings, documentation and copyright notice?
 - [x] Is this contribution tested against [all ArcticDB's features](../docs/mkdocs/docs/technical/contributing.md)?
 - [x] Do all exceptions introduced raise appropriate [error messages](https://docs.arcticdb.io/error_messages/)?
 - [x] Are API changes highlighted in the PR description?
 - [x] Is the PR labelled as enhancement or bug so it appears in autogenerated release notes?
</details>

### CODE REVIEW

**Review by JohanMabille — APPROVED** [review]

File: python/tests/nonreg/arcticdb/version_store/test_nonreg_specific.py [hunk 1]
```diff
@@ -197,3 +198,100 @@ def test_batch_write_unicode_strings(lmdb_version_store):
...
+    # Non-regression test for https://github.com/man-group/ArcticDB/issues/987
+    lib = lmdb_version_store
+
+    # Columns of numeric (which includes datetimes) dtype != float64 with no values
+    # must always be stored with the provided dtype (never empty-type).
+
+    symbol_info_series_index = 1 if IS_PANDAS_TWO else 0
+    def get_symbol_value_type(symbol_info):
+        symbol_info = lib.get_info(symbol)
+        return symbol_info["dtype"][symbol_info_series_index].value_type
```
Comment (alexowens90) [inline python/tests/nonreg/arcticdb/version_store/test_nonreg_specific.py:245]: `TypeDescriptor` also has a `size_bits` field we could check here too

File: python/tests/nonreg/arcticdb/version_store/test_nonreg_specific.py [hunk 2]
```diff
@@ -197,3 +198,100 @@ def test_batch_write_unicode_strings(lmdb_version_store):
...
+    # Non-regression test for https://github.com/man-group/ArcticDB/issues/940
+    lib = lmdb_version_store
+
+    symbol = "test_update_with_empty_dataframe_with_index"
+
+    series = pd.Series(dtype="datetime64[ns]")
+    lib.write(symbol, series)
+
+    # This must not fail.
+    lib.read(symbol, as_of=0).data
```
Comment (alexowens90) [inline python/tests/nonreg/arcticdb/version_store/test_nonreg_specific.py:232]: I'd like to make these tests more comprehensive, covering the entire matrix of:

- `write`, `append`, and `update` methods
- existing symbol is empty/has data
- dataframe being written/appended/updated is empty/has data
- static/dynamic schema

**Review by alexowens90 — CHANGES_REQUESTED** [review]

File: python/tests/nonreg/arcticdb/version_store/test_nonreg_specific.py (hunk 1 above)
Comment (jjerphan) [reply python/tests/nonreg/arcticdb/version_store/test_nonreg_specific.py]: Done in a473bae.

File: python/tests/nonreg/arcticdb/version_store/test_nonreg_specific.py (hunk 2 above)
Comment (jjerphan) [reply python/tests/nonreg/arcticdb/version_store/test_nonreg_specific.py:269]: Started with 2a8bb7a.

File: cpp/arcticdb/entity/types.hpp [hunk 3]
```diff
@@ -115,7 +115,10 @@ enum class ValueType : uint8_t {
 
     UTF_DYNAMIC = 11,
     ASCII_DYNAMIC = 12,
-    /// Used to represent null types. Each type can be converted to Empty and Empty can be converted to each type.
+    // EMPTY is used by empty columns whose type has not been explicitly specified.
```
Comment (vasil-pashov) [inline cpp/arcticdb/entity/types.hpp:118]: It's used not only for empty columns, but for columns full of `None` values as well. E.g. `pd.DataFrame({"c": [None, None, None]})` would create a column of EMPTY type in Arctic

File: cpp/arcticdb/python/python_to_tensor_frame.cpp [hunk 4]
```diff
@@ -73,7 +73,10 @@ NativeTensor obj_to_tensor(PyObject *ptr) {
     auto descr = pybind11::detail::array_descriptor_proxy(arr->descr);
     auto ndim = arr->nd;
     ssize_t size = ndim == 1 ? arr->dimensions[0] : arr->dimensions[0] * arr->dimensions[1];
-    auto val_type = size > 0 ? get_value_type(descr->kind) : ValueType::EMPTY;
+    // In Pandas < 2, empty series dtype is `"float"`, but as of Pandas 2.0, empty series dtype is `"object"`
+    // The Normalizer in Python cast empty `"float"` series to `"object"` so `EMPTY` is used here.
+    // See https://github.com/man-group/ArcticDB/issues/987#issue-1956701712
+    auto val_type = (size == 0 && descr->kind == 'O') ? ValueType::EMPTY : get_value_type(descr->kind);
```
Comment (vasil-pashov) [inline cpp/arcticdb/python/python_to_tensor_frame.cpp:79]: Does this mean that if we explicitly set the dtype of an empty column it won't have EMPTY type even though it's empty. I am not sure that this is the desired behavior. This needs unit tests.

Comment (jjerphan) [reply cpp/arcticdb/python/python_to_tensor_frame.cpp:79]: Yes, this was agreed on https://github.com/man-group/ArcticDB/issues/987 (see the description of the PR).

File: python/tests/nonreg/arcticdb/version_store/test_nonreg_specific.py [hunk 5]
```diff
@@ -197,3 +198,155 @@ def test_batch_write_unicode_strings(lmdb_version_store):
...
+
+
+def test_empty_column_handling(lmdb_version_store):
```
Comment (vasil-pashov) [inline python/tests/nonreg/arcticdb/version_store/test_nonreg_specific.py:278]: This tests seems rather long to me, can we split it or use some fixture to make it a bit more readable?

Comment (jjerphan) [reply python/tests/nonreg/arcticdb/version_store/test_nonreg_specific.py]: Done with ff99224.

File: cpp/arcticdb/entity/types.hpp (hunk 3 above)
Comment (jjerphan) [reply cpp/arcticdb/entity/types.hpp]: Done with ff99224.

**Review by alexowens90 — CHANGES_REQUESTED** [review]

**Review by alexowens90 — APPROVED** [review]

**Pull request merged**